// webuikit/src/Components/Typography/Heading.js
import { fontSizes, fontWeights, lineHeight } from "tokens/typography";
import { black } from "ZColors";
import styled from "styled-components";

export const H1 = styled.h1`
  font-size: ${props => props.fontSize || fontSizes.z900};
  line-height: ${lineHeight.regular};
  font-weight: ${fontWeights.regular};
  color: ${black};
`;
export const H2 = styled.h2`
  font-size: 1.5rem;
  font-size: ${props => props.fontSize || fontSizes.z800};
  line-height: ${lineHeight.regular};
  font-weight: ${fontWeights.regular};
  color: ${black};
`;
export const H3 = styled.h3`
  font-size: ${props => props.fontSize || fontSizes.z700};
  line-height: ${lineHeight.regular};
  font-weight: ${fontWeights.regular};
  color: ${black};
`;
export const H4 = styled.h4`
  font-size: ${props => props.fontSize || fontSizes.z600};
  line-height: ${lineHeight.regular};
  font-weight: ${fontWeights.regular};
  color: ${black};
`;
export const H5 = styled.h5`
  font-size: ${props => props.fontSize || fontSizes.z500};
  line-height: ${lineHeight.regular};
  font-weight: ${fontWeights.regular};
  color: ${black};
`;
export const H6 = styled.h6`
  font-size: ${props => props.fontSize || fontSizes.z400};
  line-height: ${lineHeight.regular};
  font-weight: ${fontWeights.regular};
  color: ${black};
`;
