// webuikit/src/Components/Slider/Slider.js
// Libraries
import React, { useState } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import InputRange from "react-input-range";
import { noop as _noop } from "lodash-es";

// styles
import "style-loader!../../../node_modules/react-input-range/lib/css/index.css";

const RangeWrapper = styled.figure`
  .input-range__track--active {
    background: #28a745;
  }

  .input-range__slider {
    height: 2rem;
    width: 2rem;
    top: -0.5rem;
    background: linear-gradient(180deg, #ffffff 0%, #f9fafb 100%);
    border: 1px solid rgba(35, 69, 97, 0.26);
    box-shadow: 0px 1px 0px rgba(22, 29, 37, 0.05);
  }

  .input-range__slider:hover {
    border: 3px solid #28a745;
  }

  .input-range__label--value .input-range__label-container {
    bottom: 1.4rem;
    left: 0;
  }
`;

const Slider = props => {
  const [value, setValue] = useState(props.value);

  const onChange = newValue => {
    props.onChange(newValue);
    setValue(newValue);
  };

  return (
    <RangeWrapper>
      <InputRange
        maxValue={props.maxValue}
        minValue={props.minValue}
        value={value}
        onChange={onChange}
      />
    </RangeWrapper>
  );
};

Slider.propTypes = {
  minValue: PropTypes.number,
  maxValue: PropTypes.number,
  value: PropTypes.object.isRequired,
  onChange: PropTypes.func
};

Slider.defaultProps = {
  minValue: 0,
  maxValue: 10,
  onChange: _noop
};

export default Slider;
