// webuikit/prod.config.js
/* eslint-disable no-undef */
/* eslint import/no-extraneous-dependencies: 0 */
const path = require("path");
const webpack = require("webpack");
const CleanWebpackPlugin = require("clean-webpack-plugin");
const env = process.env.WEBPACK_BUILD || process.env.NODE_ENV || "development";

const config = {
  mode: env,
  devtool: "source-map",
  node: {
    fs: "empty"
  },
  output: {
    filename: "index.js",
    path: path.resolve(__dirname, "dist"),
    libraryTarget: "umd"
  },
  plugins: [
    new CleanWebpackPlugin(["build"]),
    new webpack.DefinePlugin({
      "process.env.NODE_ENV": JSON.stringify(env)
    }),
    new webpack.ProgressPlugin()
  ],
  entry: {
    main: "./src/index.js"
  },
  module: {
    rules: [
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"]
      },
      {
        test: /\.json$/,
        use: "json-loader?cacheDirectory"
      },
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        use: "babel-loader?cacheDirectory"
      }
      // {
      //   test: /\.css$/,
      //   include: "/src/Components/Datepicker",
      //   use: [
      //     { loader: "style-loader" },
      //     { loader: "css-loader" },
      //     { loader: "postcss-loader" }
      //   ]
      // }
    ]
  },
  // externals: {
  //   react: "commonjs react"
  // },
  resolve: {
    extensions: ['.scss', '.js', '.json', '.png', '.gif', '.jpg', '.svg'],
    alias: {
      sauce: path.resolve("./src"),
      ZColors: path.resolve(__dirname, "./src/tokens/color"),
      WebUiKIt: path.resolve(__dirname),
      Components: path.resolve(__dirname, "./src/Components"),
      tokens: path.resolve("./src/tokens")
    }
  }
};

module.exports = config;
