// webuikit/src/Components/Flag/Flag.js
import React from "react";
import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import { getCDN } from "../../helpers/commonHelper";

const DEFAULT_WIDTH = 24;
const DEFAULT_XY = "0 0";

const FLAGS_SPRITE_URL =
  "data/webuikit/39194387e4bcd5ee03cc42eb19e6febf1556092738.png";

const Flag = props => {
  const { width = DEFAULT_WIDTH, country = "" } = props;
  const xy = getXY(country);
  const scale = getScaleFactor(DEFAULT_WIDTH, width);
  const leftMargin = getLeftMargin(scale, DEFAULT_WIDTH);
  return (
    <Flag.Image
      width={DEFAULT_WIDTH}
      xy={xy}
      scaleFactor={scale}
      leftMargin={leftMargin}
    />
  );
};

const getLeftMargin = (scale, defaultWidth) => {
  scale === 1 ? 0 : defaultWidth * scale;
  if (!scale) return 0;
  return Math.floor((defaultWidth * scale - defaultWidth) / 2);
};

const getScaleFactor = (defaultWidth, width) => width / defaultWidth;

const XY_MAPPING = {
  australia: "-2.4rem 0",
  bahamas: "-9.6rem 0",
  belgium: "-4.8rem 0",
  belize: "-12rem 0",
  brazil: "-7.2rem 0",
  canada: "0 -2.4rem",
  chile: "-2.4rem -2.4rem",
  colombia: "-4.8rem -2.4rem",
  "czech-republic": "-7.2rem -2.4rem",
  denmark: "-9.6rem -2.4rem",
  finland: "-12rem -2.4rem",
  india: "-4.8rem -4.8rem",
  indonesia: "0 -4.8rem",
  ireland: "-2.4rem -4.8rem",
  italy: "-7.2rem -4.8rem",
  lebanon: "-9.6rem -4.8rem",
  luxembourg: "-12rem -4.8rem",
  malaysia: "-2.4rem -7.2rem",
  mauritius: "0 -7.2rem",
  netherlands: "-4.8rem -7.2rem",
  "new-zealand": "-9.6rem -7.2rem",
  norway: "-7.2rem -7.2rem",
  philippines: "-12rem -7.2rem",
  poland: "0 -9.6rem",
  portugal: "-2.4rem -9.6rem"
};

const getXY = name => {
  return XY_MAPPING[name] || DEFAULT_XY;
};

Flag.propTypes = {
  width: PropTypes.number,
  country: PropTypes.string
};

Flag.Image = styled.span`
  width: ${props => props.width / 10}rem;
  height: ${props => props.width / 10}rem;
  display: inline-block;
  background: url(${getCDN(FLAGS_SPRITE_URL)}) no-repeat;
  background-position: ${props => props.xy};
  transform: scale(${props => props.scaleFactor});
  ${props =>
    !!props.leftMargin &&
    css`
      margin-left: ${props.leftMargin / 10}rem;
    `}
`;

export default Flag;
