// webuikit/src/Components/RadioGroup/RadioGroup.js
// libraries
import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { noop as _noop } from "lodash-es";
import styled from "styled-components";

// components
import Radio from "../Radio";

// helpers
import { getAttributes } from "./radiogroupHelpers";

const RadioGroup = props => {
  const [selected, setValue] = useState(props.defaultValue);

  useEffect(() => {
    props.onChange(selected);
  });

  const onChange = e => {
    setValue(e.target.value);
  };

  return (
    <RadioGroup.Element onChange={onChange} horizontal={props.horizontal}>
      {props.options.map(({ label, value, disabled }) => (
        <Radio
          key={props.name}
          name={props.name}
          value={value}
          label={label}
          checked={value === selected}
          disabled={!!disabled}
        />
      ))}
    </RadioGroup.Element>
  );
};

RadioGroup.Element = styled.div`
  display: ${getAttributes("display")};
  justify-content: ${getAttributes("justifyContent")};
`;

RadioGroup.propTypes = {
  onChange: PropTypes.func,
  horizontal: PropTypes.bool,
  defaultValue: PropTypes.string,
  name: PropTypes.string.isRequired,
  options: PropTypes.array.isRequired
};

RadioGroup.defaultProps = {
  onChange: _noop,
  horizontal: false
};

export default RadioGroup;
