// webuikit/src/Components/Footer/contents.js
export const companyContent = [
  {
    label: "About",
    link: "/about"
  },
  {
    label: "Blog",
    link: "/blog"
  },
  {
    label: "Careers",
    link: "/career"
  },
  {
    label: "Contact",
    link: "/contact"
  }
];

export const foodiesContent = [
  {
    label: "Code of conduct",
    link: "/codeofconduct"
  },
  {
    label: "Community",
    link: "/community"
  },
  {
    label: "Verified Users",
    link: "/users"
  },
  {
    label: "Blogger Help",
    link: "/bloggerhelp"
  },
  {
    label: "Developers",
    link: "/developers"
  },
  {
    label: "Mobile Apps",
    link: "/apps"
  }
];

export const restaurantsContent = [
  {
    label: "Add a Restaurant",
    link: "/codeofconduct"
  },
  {
    label: "Claim your Listing",
    link: "/community"
  },
  {
    label: "Business App",
    link: "/users"
  },
  {
    label: "Business Owner Guidelines",
    link: "/bloggerhelp"
  },
  {
    label: "Business Blog",
    link: "/developers"
  },
  {
    label: "Restaurant Widgets",
    link: "/apps"
  },
  {
    label: "Products for Businesses",
    link: "/apps"
  },
  {
    label: "Order",
    link: "/apps"
  },
  {
    label: "Book",
    link: "/apps"
  },
  {
    label: "Trace",
    link: "/apps"
  },
  {
    label: "Hyperpure",
    link: "/apps"
  }
];

export const forYouContent = [
  {
    label: "Privacy",
    link: "/codeofconduct"
  },
  {
    label: "Terms",
    link: "/community"
  },
  {
    label: "API Policy",
    link: "/users"
  },
  {
    label: "CSR",
    link: "/bloggerhelp"
  },
  {
    label: "Security",
    link: "/developers"
  },
  {
    label: "Sitemap",
    link: "/apps"
  }
];

export const icons = [
  {
    name: "facebook-circle",
    link: "/facebook"
  },
  {
    name: "twitter-circle",
    link: "/twitter"
  },
  {
    name: "instagram-circle",
    link: "/instagram"
  }
];

export const disclaimer =
  "By continuing past this page, you agree to our Terms of Service, Cookie Policy, Privacy Policy and Content Policies. All trademarks are properties of their respective owners. 2008-2019 © Zomato™ Media Pvt Ltd. All rights reserved.";

export const countries = [
  {
    name: "Australia",
    value: "australia"
  },
  {
    name: "Bahamas",
    value: "bahamas"
  },
  {
    name: "Belgium",
    value: "belgium"
  },
  {
    name: "Belize",
    value: "belize"
  },
  {
    name: "Brazil",
    value: "brazil"
  },
  {
    name: "Canada",
    value: "canada"
  },
  {
    name: "Chile",
    value: "chile"
  },
  {
    name: "Colombia",
    value: "colombia"
  },
  {
    name: "Czech Republic",
    value: "czech-republic"
  },
  {
    name: "Denmark",
    value: "denmark"
  },
  {
    name: "Finland",
    value: "finland"
  },
  {
    name: "India",
    value: "india"
  },
  {
    name: "Indonesia",
    value: "indonesia"
  },
  {
    name: "Ireland",
    value: "ireland"
  },
  {
    name: "Italy",
    value: "italy"
  },
  {
    name: "Lebanon",
    value: "lebanon"
  },
  {
    name: "Luxembourg",
    value: "luxembourg"
  },
  {
    name: "Malaysia",
    value: "malaysia"
  },
  {
    name: "Mauritius",
    value: "mauritius"
  },
  {
    name: "Netherlands",
    value: "netherlands"
  },
  {
    name: "New Zealand",
    value: "new-zealand"
  },
  {
    name: "Norway",
    value: "norway"
  },
  {
    name: "Philippines",
    value: "philippines"
  },
  {
    name: "Poland",
    value: "poland"
  },
  {
    name: "Portugal",
    value: "portugal"
  }
];

export const languages = [
  {
    name: "Čeština",
    value: "CE"
  },
  {
    name: "English",
    value: "EN"
  },
  {
    name: "Español",
    value: "SP"
  },
  {
    name: "Indonesian",
    value: "ID"
  },
  {
    name: "Italian",
    value: "IT"
  },
  {
    name: "Polish",
    value: "PO"
  },
  {
    name: "Português (BR)",
    value: "BR"
  },
  {
    name: "Português (PT)",
    value: "PT"
  },
  {
    name: "Slovenčina",
    value: "SL"
  },
  {
    name: "Türkçe",
    value: "TR"
  },
  {
    name: "Vietnamese",
    value: "VT"
  }
];
