// webuikit/src/Components/Switch/RadioSwitch.js
import React, { useState } from "react";
import styled from "styled-components";
import misc from "../../tokens/misc";
import { noop as _noop } from "lodash-es";
import PropTypes from "prop-types";
import { red, grey } from "ZColors";
import Icon from "Components/Icon";

let size = 26;
let spacing = 2;

const RadioSwitch = ({ onToggle, readOnly, active }) => {
  const [activeState, toggleActiveState] = useState(active);
  return (
    <RadioSwitch.wrapper
      onClick={() => {
        onToggle(!activeState);
        return toggleActiveState(!activeState);
      }}
    >
      <RadioSwitch.iconSpan on={activeState && !readOnly}>
        <Icon color={red.z500} size={size} name="checkCircle" />
      </RadioSwitch.iconSpan>
      <RadioSwitch.inactiveSpan on={!activeState || readOnly} />
    </RadioSwitch.wrapper>
  );
};

RadioSwitch.wrapper = styled.span`
  display: inline-block;
  position: relative;
  cursor: pointer;
  width: ${size / 10}rem;
  height: ${size / 10}rem;
  appearance: none;
  outline: none;
`;

RadioSwitch.inactiveSpan = styled.span`
  display: inline-block;
  position: absolute;
  appearance: none;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  border: ${spacing / 10}rem ${grey.z400} solid;
  border-radius: 50%;
  transition: opacity ${misc.animationDuration};
  opacity: ${props => (props.on ? "1" : 0)};
`;

RadioSwitch.iconSpan = styled.span`
  display: inline-block;
  position: absolute;
  top: 0;
  left: 0;
  width: ${size / 10}rem;
  appearance: none;
  height: ${size / 10}rem;
  transition: opacity ${misc.animationDuration};
  opacity: ${props => (props.on ? "1" : 0)};
`;

RadioSwitch.propTypes = {
  readOnly: PropTypes.bool,
  size: PropTypes.number,
  spacing: PropTypes.number,
  onToggle: PropTypes.func,
  active: PropTypes.bool
};

RadioSwitch.defaultProps = {
  readOnly: false,
  size: 26,
  spacing: 2,
  onToggle: _noop
};

export default RadioSwitch;
