// webuikit/.storybook/config.js
import {
  addDecorator,
  addParameters,
  configure,
  setAddon,
  getStorybook
} from "@storybook/react";
import JSXAddon from "storybook-addon-jsx";
import { addReadme } from "storybook-readme";
import zomaTheme from "./zomaTheme";
import GlobalStyles from "./globalstyles";
import React from "react";
import { withA11y } from "@storybook/addon-a11y";
import { withKnobs } from "@storybook/addon-knobs";
import { withInfo } from "@storybook/addon-info";
import chaptersAddon, { setDefaults } from "react-storybook-addon-chapters";
import styled from "styled-components";
import { themes } from "@storybook/theming";

setAddon(JSXAddon);
addDecorator(withA11y);
addDecorator(withKnobs);
setAddon(chaptersAddon);
addParameters({
  info: {
    inline: true,
    header: false
  },
  options: {
    theme: zomaTheme
  },
  readme: {
    // You can set the global code theme here.
    codeTheme: "a11y-dark"
  }
});
addDecorator(addReadme);
const Layout = styled.div`
  display: flex;
  padding: 2em;
  flex-direction: column;
  justify-content: center;
  align-items: space-between;
  & > * {
    max-width: 100%;
  }
`;

function loadStories() {
  const req = require.context("../stories", true, /\.stories\.(js|md)$/);
  req.keys().forEach(filename => req(filename));
}

const withGlobal = story => (
  <React.Fragment>
    <GlobalStyles />
    <Layout>{story()}</Layout>
  </React.Fragment>
);

addDecorator(addReadme);
addDecorator(withGlobal);
configure(loadStories, module);

export { getStorybook };
