// webuikit/.storybook/webpack.config.js
/* eslint import/no-extraneous-dependencies: 0 */
const path = require("path");

// Export a function. Accept the base config as the only param.
module.exports = async ({ config, mode }) => {
  // `mode` has a value of 'DEVELOPMENT' or 'PRODUCTION'
  // You can change the configuration based on that.
  // 'PRODUCTION' is used when building the static version of storybook.

  // Make whatever fine-grained changes you need
  config.module.rules.push({
    test: /\.css$/,
    loaders: [
      require.resolve("style-loader"),
      {
        loader: require.resolve("css-loader"),
        options: {
          importLoaders: 1,
          modules: true,
          localIdentName: "[name]__[local]___[hash:base64:5]"
        }
      }
    ],
    include: path.resolve(__dirname, "../src/Components/")
  });

  config.module.rules.push({
    test: /\.(js|jsx)$/,
    exclude: /node_modules/,
    use: "babel-loader"
  });

  config.module.rules.push({
    test: /\.json$/,
    exclude: /node_modules/,
    use: "json-loader"
  });

  config.resolve.alias["sauce"] = path.resolve(__dirname, "../src");
  config.resolve.alias["ZColors"] = path.resolve(
    __dirname,
    "../src/tokens/color/index.js"
  );
  config.resolve.alias["Components"] = path.resolve(
    __dirname,
    "../src/Components"
  );

  config.resolve.alias["tokens"] = path.resolve(__dirname, "../src/tokens");

  // Return the altered config
  return config;
};
