// webuikit/src/Layouts/Navbar.js
/* eslint-disable indent */
import React from "react";
import styled, { css } from "styled-components";
import { grey, red } from "ZColors";
import PropTypes from "prop-types";
import { fontSizes, fontWeights, lineHeight } from "../tokens/typography";

const Navbar = ({ children }) => <Navbar.Element>{children}</Navbar.Element>;

const activeLinkStyle = css`
  color: #151515;
  font-weight: ${fontWeights.regular};
  border-bottom-color: ${red.z500};
`;
Navbar.Element = styled.nav`
  width: 80%;
  ${"" /* max-width: 90rem; */}
  height: 100%;
  ${"" /* width: 70%; */}
  background: transparent;
  padding-top: 1.5rem;
  padding-bottom: 1.7rem;
  display: flex;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: flex-end;
  cursor: pointer;
  opacity: 1;
  transition: opacity linear 0.25s;
  padding: 1.7rem 0;
  > * {
    margin-right: 3.2rem;
  }
  > *:last-child {
    margin-right: 0;
  }
  @media screen and (max-width: 830px) {
    > * {
      margin-right: 1rem;
    }
  }

  @media screen and (max-width: 768px) {
    & {
      opacity: 0;
    }
  }
`;

Navbar.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};

Navbar.Link = styled.a`
  font-size: ${fontSizes.z400};
  line-height: ${lineHeight.medium};
  text-align: right;
  vertical-align: middle;
  font-weight: ${fontWeights.light};
  color: ${grey.z900};
  padding: 1rem 0;
  box-sizing: border-box;
  border-width: 0.2rem;
  border-style: solid;
  border-color: transparent;

  &:hover,
  &:active {
    ${activeLinkStyle}
  }

  ${props => props.active && activeLinkStyle}
`;

export default Navbar;
